src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -592,6 +617,25 @@ private void MakeExplicitParameterTypeInferences(BoundExpression argument, TypeW
                 {
                     ExactOrBoundsInference(kind, _extensions.GetTypeWithAnnotations(argument), target, ref useSiteInfo);
                 }
+                else if (IsUnfixedTypeParameter(target))
+                {
+                    var ordinal = ((TypeParameterSymbol)target.Type).Ordinal;
+                    var typeWithAnnotations = _extensions.GetTypeWithAnnotations(argument);
```

Comment:
Consider factoring the `.NullableAnnotation` up here and changing the local's name.

Reply:
I'm debating how closely we should stick to the spec. Maybe @cston could advise.

The spec says that we're recording a special bound for `null`, which flows into fixing:
"Otherwise if Ei is null then U is the special bound `null`"
"Fixing:... If the resulting candidate is a reference type or a nonnullable value type and all of the exact bounds or any of the lower bounds are nullable value types, nullable reference types, null or default, then ? is added to the resulting candidate, making it a nullable value type or reference type."

Did you try to flow a TypeWithAnnotations(type: null, NullableAnnotation.Annotated) into fixing?

Reply:
I did initially try to make a TypeWithAnnotations with a null Type work. I found that the assumption that a TypeWithAnnotations would have a type in here was baked in fairly deeply--changing it to work differently requiring many changes to asserts and additional HasType checks, etc, and that the handling that is actually needed for the "annotation-only" bounds is limited enough that just storing them on the side felt considerably simpler.

Reply:
> Consider factoring the .NullableAnnotation up here and changing the local's name.

I found that calling `_extensions.GetTypeWithAnnotations(argument)` outside of `IsReallyAType(argument.GetTypeOrFunctionType())` or `IsUnfixedTypeParameter(target)` caused us to hit the following assert in `OverloadResolutionTests.GenericInferenceOnInErrTuples`:

https://github.com/dotnet/roslyn/blob/1dd98de8b262e5dfe03edbba9858f242da32cadb/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L6133-L6157

The kind is 'ConvertedTupleLiteral' and the offending expression is `(null, 1)`. It wasn't clear to me if anything really should be adjusted in `GetNullableAnnotation()`, so I decided to just delay calling `GetTypeWithAnnotations` until closer to where it is used.